src/types/capsule.rs

```diff
@@ -254,28 +317,92 @@ impl<'py> PyCapsuleMethods<'py> for Bound<'py, PyCapsule> {
         }
     }
 
+    fn pointer_checked(&self, name: Option<&CStr>) -> PyResult<NonNull<c_void>> {
+        // SAFETY:
+        // - `self.as_ptr()` is a valid object pointer
+        // - `name_ptr` is either a valid C string or null
+        // - thread is attached to the Python interpreter
+        let ptr = unsafe { ffi::PyCapsule_GetPointer(self.as_ptr(), name_ptr(name)) };
+        match NonNull::new(ptr) {
+            Some(ptr) => Ok(ptr),
+            None => Err(PyErr::fetch(self.py())),
+        }
```

Comment:
aside: if you prefer more functional style:
```suggestion
        NonNull::new(ptr).ok_or_else(|| PyErr::fetch(self.py()))
```